File: 2019/js/03.js
let fs = require('fs')

const lines = fs.readFileSync('../inputs/03').toString().split('\n').map(line => line.split(','))

const test_inputs_1 = [
    "R75,D30,R83,U83,L12,D49,R71,U7,L72".split(","),
    "U62,R66,U55,R34,D71,R55,D58,R83".split(",")
]
const test_inputs_2 = [
    "R98,U47,R26,D63,R33,U87,L62,D20,R33,U53,R51".split(","),
    "U98,R91,D20,R16,D67,R40,U7,R15,U6,R7".split(",")
]

function generate_points(instructions) {
    let current_point = [0, 0]
    const points = []
    instructions.forEach(instruction => {
        const direction = instruction.substring(0, 1)
        const value = parseInt(instruction.substring(1))
        Array(value).fill().map((_, i) => {
            if (direction === 'R')
                current_point[1]++
            else if (direction === 'L')
                current_point[1]--
            else if (direction === 'U')
                current_point[0]++
            else if (direction === 'D')
                current_point[0]--
            points.push([...current_point])
        })
    })

    return points
}

function manhattan_distance(a, b) {
    console.log(a)
    const vals = a.map((x1, i) => {
        return Math.abs(x1 - b[i])
    })


    return vals.reduce((acc, val) => acc + val, 0)
}

function common_elements(a, b) {
    // See https://exploringjs.com/impatient-js/ch_sets.html#missing-set-operations
    const set_a = [...new Set([...a].map((xy, _) => `${xy[0]},${xy[1]}`))]
    const set_b = [...new Set([...b].map((xy, _) => `${xy[0]},${xy[1]}`))]

    const instersection = [...set_a].filter(el => set_b.indexOf(el) >= 0).map(el => el.split(",").map(s => parseInt(s)))

    // const instersection = [...set_a].filter(el => {
    //     return [...set_b].filter(innerEl => el[0] === innerEl[0] && el[1] === innerEl[1]).length > 0
    // })

    return instersection
}

function min_manhattan_distance(a, b) {
    const points_a = generate_points(a)
    const points_b = generate_points(b)
    const common_points = common_elements(points_a, points_b)
    const distances = common_points.map(point => manhattan_distance(point, [0, 0]))

    return Math.min(...distances)
}

// Example 1
console.log(`Test set 1: ${min_manhattan_distance(...test_inputs_1)}`)
// Example 2
console.log(`Test set 2: ${min_manhattan_distance(...test_inputs_2)}`)

// Result 1
console.log(`Test set 1: ${min_manhattan_distance(...lines)}`)

File: 2019/python/03.py
#!/usr/bin/env python
# coding: utf-8

# ## Part 1

# In[1]:


with open('../inputs/03', 'r') as f:
    file_lines = f.readlines()
input_lines = [[item.strip() for item in line.split(",")]
               for line in file_lines]


# In[2]:


# input_lines


# In[3]:


test_inputs_1 = [
    "R75,D30,R83,U83,L12,D49,R71,U7,L72".split(","),
    "U62,R66,U55,R34,D71,R55,D58,R83".split(",")
]
test_inputs_2 = [
    "R98,U47,R26,D63,R33,U87,L62,D20,R33,U53,R51".split(","),
    "U98,R91,D20,R16,D67,R40,U7,R15,U6,R7".split(",")
]


# In[4]:


def generate_points(instructions):
    current_point = [0, 0]
    points = []
    for instruction in instructions:
        direction = instruction[0]
        value = int(instruction[1:])
        for i in range(0, value):
            if direction == "R":
                current_point[1] += 1
            elif direction == "L":
                current_point[1] -= 1
            elif direction == "U":
                current_point[0] += 1
            elif direction == "D":
                current_point[0] -= 1
            points.append(tuple(current_point.copy()))
    return points


# In[5]:


def manhattan_distance(a, b):
    return sum([abs(x1 - x2) for x1, x2 in zip(a, b)])


# In[6]:


def common_elements(a, b):
    return set(a) & set(b)


# In[7]:


def compute_min_manhattan_distance(a, b):
    points_a = generate_points(a)
    points_b = generate_points(b)
    common_points = common_elements(points_a, points_b)
    distances = [manhattan_distance(point, [0, 0]) for point in common_points]

    return min(distances)


# In[8]:


compute_min_manhattan_distance(*test_inputs_1)


# In[9]:


compute_min_manhattan_distance(*test_inputs_2)


# In[10]:


# Result 1
compute_min_manhattan_distance(*input_lines)


# ## Part 2

# In[11]:


def step_distance(point, a, b):
    step_counter = 0
    for step in a:
        step_counter += 1
        if step == point:
            break
    for step in b:
        step_counter += 1
        if step == point:
            break
    return step_counter


# In[12]:


def compute_min_step_distance(a, b):
    points_a = generate_points(a)
    points_b = generate_points(b)
    common_points = common_elements(points_a, points_b)
    distances = [step_distance(point, points_a, points_b)
                 for point in common_points]

    return min(distances)


# In[13]:


compute_min_step_distance(*test_inputs_1)


# In[14]:


compute_min_step_distance(*test_inputs_2)


# In[15]:


# Result 2
compute_min_step_distance(*input_lines)

